### ISSUE TITLE
地图卡顿

### ISSUE THREAD

**Issue by AFoolLove** [mapeditor/tiled#3656]
我有一张大小为256*256的地图，从1.10开始，缩放的越小，使用鼠标中键拖动地图查看时，它就越卡顿。
我尝试回退到1.9.2，1.9.2不会出现这个问题

**To Reproduce**
创建一张大小为256*256的地图，使用任意图块进行填充
Ctrl+滚轮进行缩小
鼠标中键按住进行拖动地图

**Specifications:**
 - OS: Windows 10
 - Tiled Version: 1.10, 1.10.1

**Comment by eishiya** [comment]
I can reproduce this. Large maps with a lot of tiles seem to have more hitching/lag when panning zoomed out. This affects both MMB drag, and space+drag.

The main difference between 1.9.2 and 1.10 is the switch to Qt6, so perhaps it's something to do with that.

**Comment by bjorn** [comment]
> The main difference between 1.9.2 and 1.10 is the switch to Qt6, so perhaps it's something to do with that.

In that case, it would be interesting to know whether the problem is also reproducible with the 32-bit Windows builds, which are still using Qt 5.15.2. They are the "Windows 7-8" builds at https://github.com/mapeditor/tiled/releases/tag/v1.10.1.

**Comment by firstmiddlelast** [comment]
Windows 7, Tiled 1.10.0 here. I reproduced the steps : 
create a 256x256 map, fill it with the bucket tool with one tile, zoom out with ctrl-wheel to 1%, and I can drag the map with the middle button without any freeze or slow.. except small ones (and high cpu usage) when I drag it outside of the viewport (I mean, when part of it is "hidden" behind other parts of the app like the toolbar on the top of the screen, or the properties panel on the left) : small freezes. 
I tried with a 1024x1024 window, same result except longer freezes when part of the map is out of sight (1-2s). 
In both cases, if the map stays fully visible, no noticeable freeze / cpu usage peak.

**Comment by bjorn** [comment]
> In both cases, if the map stays fully visible, no noticeable freeze / cpu usage peak.

This is normal, because in this case, the visible map is simply moved on the screen, instead of being re-drawn. This is an optimization, which is not possible when moving the camera makes a previously invisible part of the map visible.

But in that regard, nothing should have changed between 1.9 and 1.10.

**Comment by cannister-of-sparrows** [comment]
```diff
diff --git a/src/tiled/tilelayeritem.cpp b/src/tiled/tilelayeritem.cpp
index b9688986..3c476138 100644
--- a/src/tiled/tilelayeritem.cpp
+++ b/src/tiled/tilelayeritem.cpp
@@ -33,6 +33,7 @@ TileLayerItem::TileLayerItem(TileLayer *layer, MapDocument *mapDocument, QGraphi
     , mMapDocument(mapDocument)
 {
     setFlag(QGraphicsItem::ItemUsesExtendedStyleOption);
+    setCacheMode(QGraphicsItem::DeviceCoordinateCache);
 
     syncWithTileLayer();
 }
```

**Comment by bjorn** [comment]
@cannister-of-sparrows I don't think we can introduce that kind of cache. It introduces a cache per-layer, which will have significant memory overhead in case of many layers or for large layers. Being in device coordinates, the cache will get also very large when zooming in.

For me it makes Tiled extremely laggy even just while hovering a 1x1 brush over a 200x200 map with more than 2 tile layers. Apparently, since the caches are so large, they are continuously thrown away, causing Qt to end up always repainting the entire tile layer. For this to work, we'd need to find a way to enable the caching only at certain zoom levels, when most of the layer is visible anyway.

Such a per-layer cache would also need to be explicitly invalidated in case of tile animation playback (by triggering `update` call on the relevant `TileLayerItem` instances).

**Comment by cannister-of-sparrows** [comment]
That's fair. I have a performant machine and this cache resolves the issue as reported on my end. If I have more time I'll investigate further as the originally reported lagging is unexpected.
